### ISSUE TITLE
Docker Container resource not idempotent with network_aliases defined

### ISSUE THREAD

**Issue by isuftin** [chef-cookbooks/docker#998]
### Cookbook version
4.2.0

### Chef-client version
14.2.0

### Platform Details
CentOS 7 VM via Test Kitchen

### Scenario:
Set network alias on container via docker_container resource and not have it change during each Chef run

### Steps to Reproduce:
My test kitchen:
```
---
driver:
  name: vagrant
  provision: true
  customize:
    cpus: 1
    memory: 1024

provisioner:
  name: chef_zero

platforms:
  - name: centos-7

verifier:
  name: inspec

suites:
  - name: default
    run_list:
      - recipe[owi_docker::test_alias]

```

my test_alias recipe in my owi_docker cookbook:
```
docker_service 'default' do
  install_method 'package'
  version '18.03.1'
  ip_forward true
  ipv4_forward true
  ipv6_forward true
  package_name 'docker-ce'
  storage_driver 'overlay2'
  storage_opts [ "overlay2.override_kernel_check=true" ]
  mount_flags "private"
end

docker_network 'default_bridge'

docker_image 'nginx' do
  repo 'nginx'
  tag '1.13.8'
  action :pull
end

docker_container 'nginx' do
  repo 'nginx'
  tag '1.13.8'
  port [ '80:80', '443:443' ]
  network_mode 'default_bridge'
  network_aliases [ 'nginx' ]
  action :run
end

```

After the initial test kitchen run, I expect that subsequent runs will be idempotent.

Run 1 (after initial rest kitchen run that sets up all service, image pulls and running container)
```
 kitchen converge
-----> Starting Kitchen (v1.21.2)
-----> Converging <default-centos-7>...
       Preparing files for transfer
       Preparing dna.json
       Resolving cookbook dependencies with Berkshelf 7.0.3...
       Removing non-cookbook files before transfer
       Preparing validation.pem
       Preparing client.rb
-----> Chef Omnibus installation detected (install only if missing)
       Transferring files to <default-centos-7>
       Starting Chef Client, version 14.2.0
       resolving cookbooks for run list: ["owi_docker::test_alias"]
       Synchronizing Cookbooks:
         - owi_docker (0.0.13)
         - ohai (5.2.2)
         - docker (4.2.0)
       Installing Cookbook Gems:
       Compiling Cookbooks...
       Converging 4 resources
       Recipe: owi_docker::test_alias
         * docker_service[default] action create
           * docker_installation_package[default] action create
             * yum_repository[Docker] action create
        * template[/etc/yum.repos.d/Docker.repo] action create (up to date)
        * execute[yum clean metadata Docker] action nothing (skipped due to action :nothing)
        * execute[yum-makecache-Docker] action nothing (skipped due to action :nothing)
        * ruby_block[package-cache-reload-Docker] action nothing (skipped due to action :nothing)
         (up to date)
             * yum_package[docker-ce] action install (up to date)
       (up to date)
            (up to date)
         * docker_network[default_bridge] action create (up to date)
         * docker_image[nginx] action pull (up to date)
         * docker_container[nginx] action run
           - stopping nginx
           - deleting nginx
           - update nginx
           -   set network_aliases to ["nginx"] (was [])
           - starting nginx

       Running handlers:
       Running handlers complete
       Chef Client finished, 1/11 resources updated in 09 seconds
       Downloading files from <default-centos-7>
       Finished converging <default-centos-7> (0m12.65s).
```

Run 2:
```
 kitchen converge
-----> Starting Kitchen (v1.21.2)
-----> Converging <default-centos-7>...
       Preparing files for transfer
       Preparing dna.json
       Resolving cookbook dependencies with Berkshelf 7.0.3...
       Removing non-cookbook files before transfer
       Preparing validation.pem
       Preparing client.rb
-----> Chef Omnibus installation detected (install only if missing)
       Transferring files to <default-centos-7>
       Starting Chef Client, version 14.2.0
       resolving cookbooks for run list: ["owi_docker::test_alias"]
       Synchronizing Cookbooks:
         - owi_docker (0.0.13)
         - docker (4.2.0)
         - ohai (5.2.2)
       Installing Cookbook Gems:
       Compiling Cookbooks...
       Converging 4 resources
       Recipe: owi_docker::test_alias
         * docker_service[default] action create
           * docker_installation_package[default] action create
             * yum_repository[Docker] action create
        * template[/etc/yum.repos.d/Docker.repo] action create (up to date)
        * execute[yum clean metadata Docker] action nothing (skipped due to action :nothing)
        * execute[yum-makecache-Docker] action nothing (skipped due to action :nothing)
        * ruby_block[package-cache-reload-Docker] action nothing (skipped due to action :nothing)
         (up to date)
             * yum_package[docker-ce] action install (up to date)
       (up to date)
            (up to date)
         * docker_network[default_bridge] action create (up to date)
         * docker_image[nginx] action pull (up to date)
         * docker_container[nginx] action run
           - stopping nginx
           - deleting nginx
           - update nginx
           -   set network_aliases to ["nginx"] (was [])
           - starting nginx

       Running handlers:
       Running handlers complete
       Chef Client finished, 1/11 resources updated in 08 seconds
       Downloading files from <default-centos-7>
       Finished converging <default-centos-7> (0m11.72s).
```

### Expected Result:
I expect the docker_container resource to not update the currently running resource as no changes have been made. 

### Actual Result:
network_alias attribute keeps changing from [] to ['nginx'] on every run. 

When I am in the VM and inspecting the container itself, it's running with the proper network alias

```
$ docker inspect nginx

[...]
            "Networks": {
                "default_bridge": {
                    "IPAMConfig": {},
                    "Links": null,
                    "Aliases": [
                        "nginx",
                        "e49161134b36"
                    ],
                    "NetworkID": "b4ec3b4e3630a3f6574e7cc512e01b7dcd2691416c84001a232f504c37407b7f",
                    "EndpointID": "fcea66d0b146409fc492c8de9ec07b2c1dbc478af4b50f78fbddf2c24008c128",
                    "Gateway": "172.18.0.1",
                    "IPAddress": "172.18.0.2",
                    "IPPrefixLen": 16,
                    "IPv6Gateway": "",
                    "GlobalIPv6Address": "",
                    "GlobalIPv6PrefixLen": 0,
                    "MacAddress": "02:42:ac:12:00:02",
                    "DriverOpts": null
                }
            }
[...]

**Comment by isuftin** [comment]
The issue seems to be that the docker-api gem hardcodes the Docker API version to 1.16.

With version 1.16, NetwrokSettings => Networks does not exist

That is where the aliases would lie.

That seems to only come in at API version 1.21.

A few problems:

First, the docker-api gem hard-codes 1.16 as the API version [by design](https://github.com/swipely/docker-api/issues/441). You can override that. For example, one can include:

```
Docker.send(:remove_const, 'API_VERSION')
Docker.const_set('API_VERSION', '1.21')
```

Right before this line...
https://github.com/chef-cookbooks/docker/blob/v4.2.0/libraries/docker_container.rb#L400

However, when you do that, the error from the Docker cookbook then becomes:

```
           ================================================================================
           Error executing action `run` on resource 'docker_container[nginx2]'
           ================================================================================

           NoMethodError
           -------------
           undefined method `new_resource' for DockerCookbook::DockerContainer

           Cookbook Trace:
           ---------------
           /tmp/kitchen/cache/cookbooks/docker/libraries/docker_container.rb:429:in `block in <class:DockerContainer>'

           Resource Declaration:
           ---------------------
           # In /tmp/kitchen/cache/cookbooks/owi_docker/recipes/test_alias.rb

            21: docker_container 'nginx2' do
            22:   repo 'nginx'
            23:   tag '1.13.8'
            24:   port [ '8080:80', '4434:443' ]
            25:   network_mode 'default_bridge'
            26:   network_aliases [ 'nginx2' ]
            27:   action :run
            28: end
            29:
```

My assumption here is that this is because the 1.21 API updates the format of the hash coming back from the Docker engine and it is incompatible with how the Docker cookbook is currently written.

In any case, hacking in the API version into the docker-api gem is..well..hacky. The docker-api gem authors should update their gem accordingly so that we can attain the most current API version output. 

Until that happens, I don't know that the docker_container resource will ever be idempotent with the stuff that is put under the `Networks` section.

Thoughts?

**Comment by isuftin** [comment]
More detailed info here: https://docs.docker.com/engine/api/version-history/#v121-api-changes

**Comment by turtleDev** [comment]
ran into this issue today. This is a big problem for me because I can't have my services restart on every periodic docker run (we're using chef-solo, and we periodically run it)

**Comment by dheerajd-msys** [comment]
Closing as it's fixed in #1042
